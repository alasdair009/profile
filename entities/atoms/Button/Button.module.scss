@import "../../../app/styles/animation";
@import "../../../app/styles/colors";
@import "../../../app/styles/typography";
@import "../../../app/styles/helpers/rem";

$buttonClipPath: polygon(
  calc(0% + #{toRem(map-get($sizes, s8))}) 0,
  100% 0,
  calc(100% - #{toRem(map-get($sizes, s8))}) 100%,
  0 100%
);

.root {
  align-items: center;
  appearance: none;
  background: transparent;
  border: none;
  color: $colorWhiteGhost;
  cursor: pointer;
  display: flex;
  font-size: toRem(map-get($fontSizes, medium));
  font-weight: map-get($fontWeights, bold);
  height: toRem(map-get($sizes, s48));
  justify-content: center;
  max-width: toRem(map-get($sizes, s256));
  min-width: toRem(map-get($sizes, s64));
  padding: 0 toRem(map-get($sizes, s48));
  position: relative;
  transition: padding animationDuration(0.5);
  z-index: 0;

  &::before,
  &::after {
    background: $colorGreenGrass;
    clip-path: $buttonClipPath;
    content: "";
    display: block;
    height: 100%;
    left: 50%;
    position: absolute;
    top: 0;
    transform: translateX(-50%);
    transition: all animationDuration(0.5);
    width: calc(100% - #{toRem(map-get($sizes, s32))});
    z-index: -1;
  }

  &::after {
    background: $colorGreyDark;
    left: calc(50% + #{toRem(5)});
    top: toRem(5);
    z-index: -2;
  }

  &:hover {
    &::before,
    &::after {
      width: calc(100% - #{toRem(map-get($sizes, s8))});
    }

    &::after {
      background: $colorWhiteGhost;
    }
  }

  &:active::before {
    background: darken($colorGreenGrass, 0.05);
  }

  &Transparent {
    height: auto;
    min-width: 0;
    padding: 0;

    &::before,
    &::after {
      display: none;
    }
  }
}
