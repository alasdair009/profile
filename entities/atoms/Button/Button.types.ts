import { HTMLAttributes, PointerEvent } from "react";

export type ButtonVariant = "standard" | "transparent";

export type ButtonProps = {
  /**
   * Action to perform when the button is clicked.
   */
  onClick?: (e: PointerEvent<HTMLButtonElement>) => void;
  /**
   * HTML Button type.
   */
  type: HTMLButtonElement["type"];
  /**
   * Visual variant of the button.
   */
  variant?: ButtonVariant;
} & HTMLAttributes<HTMLButtonElement>;
