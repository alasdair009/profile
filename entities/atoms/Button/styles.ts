"use client";
import styled from "styled-components";
import { colors } from "../../design-tokens/colors";
import { sizes } from "../../design-tokens/dimensions";
import { darken, rem } from "polished";
import { fontSizes, fontWeights } from "../../design-tokens/typography";

export const Root = styled.button`
  align-items: center;
  appearance: none;
  background: ${colors.greenGrass};
  border: ${rem(1)} solid ${darken(0.25, colors.greenGrass)};
  color: ${colors.whiteGhost};
  cursor: pointer;
  display: flex;
  font-size: ${fontSizes.medium.rem};
  font-weight: ${fontWeights.bold};
  height: ${sizes.s48.rem};
  justify-content: center;
  min-width: ${sizes.s64.rem};
  max-width: ${sizes.s256.rem};
  padding: 0 ${sizes.s16.rem};

  &:hover {
    filter: brightness(1.3);
  }

  &:active {
    background: ${darken(0.05, colors.greenGrass)};
  }
`;
