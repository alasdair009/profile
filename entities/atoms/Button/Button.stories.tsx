import { Button } from "./Button";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof Button> = {
  component: Button,
};
export default meta;

export const Default: StoryObj<typeof Button> = {};
