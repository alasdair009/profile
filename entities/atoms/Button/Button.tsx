import { Root } from "./styles";
import { ButtonProps } from "@/entities/atoms/Button/Button.types";

/**
 * Clickable button for executing actions
 */
export function Button({ variant = "standard", ...rest }: ButtonProps) {
  return <Root $variant={variant} {...rest} />;
}
