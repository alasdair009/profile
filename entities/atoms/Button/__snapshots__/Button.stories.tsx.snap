// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/Button Default play-test 1`] = `
<button data-testid="Button"
        class="sc-hDcvty jPmyol"
>
  Button
</button>
`;
